File: 1590. Make Sum Divisible by P/solution.js
/**
 * @param {number[]} nums
 * @param {number} p
 * @return {number}
 */
var minSubarray = function(nums, p) {
    let targetSum = nums.reduce((acc, num) => acc + num, 0) % p;
    const n = nums.length;
    let minLen = n, currSum = 0;

    if (targetSum === 0) {
        return 0;
    }

    const remainderIdx = new Map();
    remainderIdx.set(0, -1);

    for (let i = 0; i < n; i++) {
        currSum = (currSum + nums[i]) % p;
        let diff = (currSum - targetSum + p) % p;

        if (remainderIdx.has(diff)) {
            minLen = Math.min(minLen, i - remainderIdx.get(diff));
        }

        remainderIdx.set(currSum, i);
    }

    return minLen === n ? -1 : minLen;
};

File: 1590. Make Sum Divisible by P/solution.cpp
class Solution {
public:
    int minSubarray(vector<int>& nums, int p) {

        int targetSum  = accumulate(begin(nums),end(nums),0L)%p,n = nums.size() , minLen = n , currSum = 0;

        if(targetSum == 0)
            return 0;

        unordered_map<int,int>remainder_idx;
        remainder_idx[0] = -1;

        for(int i=0;i<n;i++){
            currSum = (currSum + nums[i])%p;
            int diff = (currSum - targetSum + p)%p;

            if(remainder_idx.count(diff)>0)
                minLen = min(minLen,i-remainder_idx[diff]);

            remainder_idx[currSum] = i;
        }
        return minLen == n ? -1 : minLen;   
    }
};